File: src/carNormalizer.ts
import fuzzysort from 'fuzzysort';
import { loadCarCatalog } from './carCatalog.js';
import type { CatalogEntry } from './types.js';

// =========================================================
// BRAND CANONICAL MAPPING
// =========================================================
export const BRAND_MAPPING: Record<string, string> = {
  // Standard Make Names
  'BMW': 'BMW',
  'MERCEDES': 'Mercedes-Benz',
  'MERCEDES-BENZ': 'Mercedes-Benz',
  'MERCEDES BENZ': 'Mercedes-Benz',
  'МЕРСЕДЕС': 'Mercedes-Benz',
  'TOYOTA': 'Toyota',
  'ТОЙОТА': 'Toyota',
  'LEXUS': 'Lexus',
  'ЛЕКСУС': 'Lexus',
  'HYUNDAI': 'Hyundai',
  'ХУНДАЙ': 'Hyundai',
  'KIA': 'Kia',
  'КИА': 'Kia',
  'CHEVROLET': 'Chevrolet',
  'ШЕВРОЛЕ': 'Chevrolet',
  'AUDI': 'Audi',
  'АУДИ': 'Audi',
  'PORSCHE': 'Porsche',
  'ПОРШ': 'Porsche',
  'HONDA': 'Honda',
  'ХОНДА': 'Honda',
  'NISSAN': 'Nissan',
  'НИССАН': 'Nissan',
  'INFINITI': 'Infiniti',
  'ИНФИНИТИ': 'Infiniti',
  'MAZDA': 'Mazda',
  'МАЗДА': 'Mazda',
  'MITSUBISHI': 'Mitsubishi',
  'МИТСУБИСИ': 'Mitsubishi',
  'VOLKSWAGEN': 'Volkswagen',
  'ФОЛЬКСВАГЕН': 'Volkswagen',
  'VW': 'Volkswagen',
  'VOLVO': 'Volvo',
  'ВОЛЬВО': 'Volvo',
  'FORD': 'Ford',
  'ФОРД': 'Ford',
  'TESLA': 'Tesla',
  'ТЕСЛА': 'Tesla',
  'BYD': 'BYD',
  'БИД': 'BYD',
  'ZEEKR': 'Zeekr',
  'ЗИКР': 'Zeekr',
  'GEELY': 'Geely',
  'ДЖИЛИ': 'Geely',
  'CHERY': 'Chery',
  'ЧЕРИ': 'Chery',
  'HAVAL': 'Haval',
  'ХАВАЛ': 'Haval',
  'EXEED': 'Exeed',
  'ЭКСИД': 'Exeed',
  'CHANGAN': 'Changan',
  'ЧАНГАН': 'Changan',
  'HONGQI': 'Hongqi',
  'ХОНЧИ': 'Hongqi',
  'LI AUTO': 'Li Auto',
  'LIXIANG': 'Li Auto',
  'ЛИ АВТО': 'Li Auto',
  'LAND ROVER': 'Land Rover',
  'RANGE ROVER': 'Land Rover',
  'RENGE ROVER': 'Land Rover',
  'ЛЕНД РОВЕР': 'Land Rover',
  'РЕЙНДЖ РОВЕР': 'Land Rover',
  'GENESIS': 'Genesis',
  'ДЖЕНЕЗИС': 'Genesis',
  'GAC': 'GAC',
  'JETOUR': 'Jetour',
  'ДЖЕТУР': 'Jetour',
  'OMODA': 'Omoda',
  'JAECOO': 'Jaecoo',
  'VOYAH': 'Voyah',
  'NIO': 'NIO',
  'XPENG': 'XPeng',
  'DENZA': 'Denza',
  'AVATR': 'Avatr',
  'TANK': 'Tank',
  'ТАНК': 'Tank',
  'ROX': 'Rox',
  'MG': 'MG',
  'BENTLEY': 'Bentley',
  'FERRARI': 'Ferrari',
  'LAMBORGHINI': 'Lamborghini',
  'ROLLS-ROYCE': 'Rolls-Royce',
  'ROLLS ROYCE': 'Rolls-Royce',
  'ASTON MARTIN': 'Aston Martin',

  // Popular Model-First & Shorthand Aliases
  'LC PRADO': 'Toyota',
  'LC PRAD0': 'Toyota',
  'LAND CRUISER PRADO': 'Toyota',
  'LC300': 'Toyota',
  'LC 300': 'Toyota',
  'LC200': 'Toyota',
  'LC 200': 'Toyota',
  'CAMRY': 'Toyota',
  'CAMRY-6': 'Toyota',
  'COROLLA': 'Toyota',
  'HIGHLANDER': 'Toyota',
  'AVALON': 'Toyota',

  'RR DEFENDER': 'Land Rover',
  'DEFENDER': 'Land Rover',
  'RR SPORT': 'Land Rover',
  'RR VOGUE': 'Land Rover',
  'RR VELAR': 'Land Rover',
  'RR AUTOBIOGRAPHY': 'Land Rover',
  'RR': 'Land Rover',

  'SANTAFE': 'Hyundai',
  'SANTA FE': 'Hyundai',
  'SANTAFEE': 'Hyundai',
  'PALISADE': 'Hyundai',

  'GELIK': 'Mercedes-Benz',
  'GELANDEWAGEN': 'Mercedes-Benz',
  'G63': 'Mercedes-Benz',
  'G 63': 'Mercedes-Benz',
  'G500': 'Mercedes-Benz',
  'G 500': 'Mercedes-Benz',
  'MAYBACH': 'Mercedes-Benz',

  'COBALT': 'Chevrolet',
  'MALIBU': 'Chevrolet',
  'MALIBU 2': 'Chevrolet',
  'GENTRA': 'Chevrolet',
  'NEXIA': 'Chevrolet',
};

export function normalizeBrandName(brand: string | null): string | null {
  if (!brand) return null;
  const cleaned = brand.replace(/\s+/g, ' ').trim();
  const upper = cleaned.toUpperCase();

  if (BRAND_MAPPING[upper]) {
    return BRAND_MAPPING[upper];
  }

  if (['BMW', 'BYD', 'NIO', 'GAC', 'MG', 'VW'].includes(upper)) {
    return upper;
  }

  return cleaned.charAt(0).toUpperCase() + cleaned.slice(1).toLowerCase();
}

export class CarCatalogNormalizer {
  private aliasMap = new Map<string, { brand: string; model: string }>();
  private searchTargets: Array<{ text: string; brand: string; model: string }> = [];

  constructor() {
    this.initCatalog();
  }

  private initCatalog(): void {
    const catalog: CatalogEntry[] = loadCarCatalog();

    for (const item of catalog) {
      const fullCanonical = `${item.brand} ${item.model}`.toUpperCase();
      this.aliasMap.set(fullCanonical, { brand: item.brand, model: item.model });
      this.searchTargets.push({ text: fullCanonical, brand: item.brand, model: item.model });

      for (const alias of item.aliases || []) {
        const upperAlias = alias.trim().toUpperCase();
        if (upperAlias) {
          this.aliasMap.set(upperAlias, { brand: item.brand, model: item.model });
          this.searchTargets.push({ text: upperAlias, brand: item.brand, model: item.model });
        }
      }
    }
  }

  public normalize(
    rawBrand: string | null,
    rawModel: string | null
  ): [string | null, string | null] {
    const normBrand = normalizeBrandName(rawBrand);
    let cleanModel = rawModel ? rawModel.replace(/\s+/g, ' ').trim() : null;

    const candidates: string[] = [];
    const words = cleanModel ? cleanModel.split(' ') : [];

    if (rawBrand && cleanModel) {
      candidates.push(`${rawBrand} ${cleanModel}`.toUpperCase());
    }
    if (cleanModel) {
      candidates.push(cleanModel.toUpperCase());
    }

    if (words.length > 0) {
      for (let i = 1; i <= Math.min(words.length, 3); i++) {
        const sub = words.slice(0, i).join(' ').toUpperCase();
        if (rawBrand) {
          candidates.push(`${rawBrand} ${sub}`.toUpperCase());
        }
        candidates.push(sub);
      }
    }

    if (rawBrand) {
      candidates.push(rawBrand.toUpperCase());
    }

    // 1. Exact lookup
    for (const cand of candidates) {
      if (this.aliasMap.has(cand)) {
        const match = this.aliasMap.get(cand)!;
        return [match.brand, match.model];
      }
    }

    // 2. Fuzzy search
    for (const cand of candidates) {
      const results = fuzzysort.go(cand, this.searchTargets, {
        key: 'text',
        threshold: -100,
        limit: 1,
      });

      if (results.length > 0 && results[0].score > -15) {
        return [results[0].obj.brand, results[0].obj.model];
      }
    }

    // Fallback for Range Rover / Land Rover
    if (rawBrand && ['RANGE ROVER', 'RENGE ROVER'].includes(rawBrand.toUpperCase())) {
      const b = 'Land Rover';
      let m = cleanModel;
      if (cleanModel && !cleanModel.toLowerCase().startsWith('range rover')) {
        m = `Range Rover ${cleanModel}`.trim();
      } else if (!cleanModel) {
        m = 'Range Rover';
      }
      return [b, m];
    }

    return [normBrand, cleanModel];
  }
}

const normalizer = new CarCatalogNormalizer();

export function normalizeCarModel(
  brand: string | null,
  model: string | null
): [string | null, string | null] {
  return normalizer.normalize(brand, model);
}

File: car_normalizer.py
import json
import re
from typing import Optional, Tuple, List, Dict
from rapidfuzz import fuzz, process

from config import CAR_CATALOG_PATH, FUZZY_MATCH_THRESHOLD

# =========================================================
# BRAND CANONICAL MAPPING
# =========================================================
BRAND_MAPPING = {
    "BMW": "BMW",
    "MERCEDES": "Mercedes-Benz",
    "MERCEDES-BENZ": "Mercedes-Benz",
    "MERCEDES BENZ": "Mercedes-Benz",
    "МЕРСЕДЕС": "Mercedes-Benz",
    "TOYOTA": "Toyota",
    "ТОЙОТА": "Toyota",
    "LEXUS": "Lexus",
    "ЛЕКСУС": "Lexus",
    "HYUNDAI": "Hyundai",
    "ХУНДАЙ": "Hyundai",
    "KIA": "Kia",
    "КИА": "Kia",
    "CHEVROLET": "Chevrolet",
    "ШЕВРОЛЕ": "Chevrolet",
    "AUDI": "Audi",
    "АУДИ": "Audi",
    "PORSCHE": "Porsche",
    "ПОРШ": "Porsche",
    "HONDA": "Honda",
    "ХОНДА": "Honda",
    "NISSAN": "Nissan",
    "НИССАН": "Nissan",
    "INFINITI": "Infiniti",
    "ИНФИНИТИ": "Infiniti",
    "MAZDA": "Mazda",
    "МАЗДА": "Mazda",
    "MITSUBISHI": "Mitsubishi",
    "МИТСУБИСИ": "Mitsubishi",
    "VOLKSWAGEN": "Volkswagen",
    "ФОЛЬКСВАГЕН": "Volkswagen",
    "VW": "Volkswagen",
    "VOLVO": "Volvo",
    "ВОЛЬВО": "Volvo",
    "FORD": "Ford",
    "ФОРД": "Ford",
    "TESLA": "Tesla",
    "ТЕСЛА": "Tesla",
    "BYD": "BYD",
    "БИД": "BYD",
    "ZEEKR": "Zeekr",
    "ЗИКР": "Zeekr",
    "GEELY": "Geely",
    "ДЖИЛИ": "Geely",
    "CHERY": "Chery",
    "ЧЕРИ": "Chery",
    "HAVAL": "Haval",
    "ХАВАЛ": "Haval",
    "EXEED": "Exeed",
    "ЭКСИД": "Exeed",
    "CHANGAN": "Changan",
    "ЧАНГАН": "Changan",
    "HONGQI": "Hongqi",
    "ХОНЧИ": "Hongqi",
    "LI AUTO": "Li Auto",
    "LIXIANG": "Li Auto",
    "ЛИ АВТО": "Li Auto",
    "LAND ROVER": "Land Rover",
    "RANGE ROVER": "Land Rover",
    "RENGE ROVER": "Land Rover",
    "ЛЕНД РОВЕР": "Land Rover",
    "РЕЙНДЖ РОВЕР": "Land Rover",
    "GENESIS": "Genesis",
    "ДЖЕНЕЗИС": "Genesis",
    "GAC": "GAC",
    "JETOUR": "Jetour",
    "ДЖЕТУР": "Jetour",
    "OMODA": "Omoda",
    "JAECOO": "Jaecoo",
    "VOYAH": "Voyah",
    "NIO": "NIO",
    "XPENG": "XPeng",
    "DENZA": "Denza",
    "AVATR": "Avatr",
    "TANK": "Tank",
    "ТАНК": "Tank",
    "ROX": "Rox",
    "MG": "MG",
    "BENTLEY": "Bentley",
    "FERRARI": "Ferrari",
    "LAMBORGHINI": "Lamborghini",
    "ROLLS-ROYCE": "Rolls-Royce",
    "ROLLS ROYCE": "Rolls-Royce",
    "ASTON MARTIN": "Aston Martin",
}


def normalize_brand_name(brand: Optional[str]) -> Optional[str]:
    """Standardizes brand string based on canonical rules."""
    if not brand:
        return None

    cleaned = re.sub(r"\s+", " ", brand).strip()
    upper = cleaned.upper()

    if upper in BRAND_MAPPING:
        return BRAND_MAPPING[upper]

    if upper == "BMW":
        return "BMW"
    if upper == "BYD":
        return "BYD"
    if upper in ("NIO", "GAC", "MG", "VW"):
        return upper

    return cleaned.title()


class CarCatalogNormalizer:
    def __init__(self, catalog_path=CAR_CATALOG_PATH):
        self.catalog: List[Dict] = []
        self.alias_to_entry: Dict[str, Tuple[str, str]] = {}
        self.alias_list: List[str] = []
        self.load_catalog(catalog_path)

    def load_catalog(self, path):
        if not path or not path.exists():
            return

        try:
            with open(path, "r", encoding="utf-8") as f:
                self.catalog = json.load(f)

            for item in self.catalog:
                brand = item.get("brand", "")
                model = item.get("model", "")
                aliases = item.get("aliases", [])

                full_canonical = f"{brand} {model}".upper()
                self.alias_to_entry[full_canonical] = (brand, model)
                if full_canonical not in self.alias_list:
                    self.alias_list.append(full_canonical)

                for alias in aliases:
                    alias_upper = alias.strip().upper()
                    if alias_upper:
                        self.alias_to_entry[alias_upper] = (brand, model)
                        if alias_upper not in self.alias_list:
                            self.alias_list.append(alias_upper)

        except Exception as e:
            print(f"⚠️ Error loading car catalog: {e}")

    def normalize(
        self,
        raw_brand: Optional[str],
        raw_model: Optional[str],
        threshold: float = FUZZY_MATCH_THRESHOLD,
    ) -> Tuple[Optional[str], Optional[str]]:
        """
        Normalizes brand and model.
        Uses RapidFuzz to match against catalog aliases.
        If fuzzy similarity >= threshold, returns normalized (brand, model).
        Otherwise falls back to cleanly normalized parsed brand and model.
        """
        norm_brand = normalize_brand_name(raw_brand)
        clean_model = re.sub(r"\s+", " ", raw_model or "").strip() if raw_model else None

        # Build candidate query strings (full and sub-phrases)
        candidates = []
        words = clean_model.split() if clean_model else []

        if raw_brand and clean_model:
            candidates.append(f"{raw_brand} {clean_model}".upper())
        if clean_model:
            candidates.append(clean_model.upper())

        # Sub-phrases if model has multiple words (e.g. "M6 4.4 V8 COMPETITION" -> "M6", "M6 4.4")
        if words:
            for i in range(1, min(len(words), 4)):
                sub = " ".join(words[:i]).upper()
                if raw_brand:
                    candidates.append(f"{raw_brand} {sub}".upper())
                candidates.append(sub)

        if raw_brand:
            candidates.append(raw_brand.upper())

        # 1. Exact lookup first
        for cand in candidates:
            if cand in self.alias_to_entry:
                b, m = self.alias_to_entry[cand]
                return b, m

        # 2. RapidFuzz matching if catalog is available
        if self.alias_list:
            best_match = None
            best_score = 0.0

            for cand in candidates:
                # token_sort_ratio for robust token matching
                match = process.extractOne(
                    cand,
                    self.alias_list,
                    scorer=fuzz.token_sort_ratio,
                )
                if match:
                    matched_alias, score, _ = match
                    if score > best_score:
                        best_score = score
                        best_match = matched_alias

            if best_match and best_score >= threshold:
                matched_brand, matched_model = self.alias_to_entry[best_match]
                return matched_brand, matched_model

        # Fallback to normalized brand and clean model
        if raw_brand and raw_brand.upper() in ("RANGE ROVER", "RENGE ROVER"):
            norm_brand = "Land Rover"
            if clean_model and not clean_model.lower().startswith("range rover"):
                clean_model = f"Range Rover {clean_model}".strip()
            elif not clean_model:
                clean_model = "Range Rover"

        return norm_brand, clean_model


# Singleton instance
_normalizer = CarCatalogNormalizer()


def normalize_car_model(
    brand: Optional[str],
    model: Optional[str],
    threshold: float = FUZZY_MATCH_THRESHOLD,
) -> Tuple[Optional[str], Optional[str]]:
    return _normalizer.normalize(brand, model, threshold=threshold)
